eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs

```diff
@@ -258,28 +268,40 @@ checkVotesAreNotForExpiredActions ::
   EpochNo ->
   [(Voter (EraCrypto era), GovActionState era)] ->
   Test (ConwayGovPredFailure era)
-checkVotesAreNotForExpiredActions curEpoch voters =
-  let votesOnExpiredActions = filter (\(_, GovActionState {gasExpiresAfter}) -> curEpoch > gasExpiresAfter) voters
-   in failureOnNonEmpty votesOnExpiredActions $
-        VotingOnExpiredGovAction . fmap (second gasId)
+checkVotesAreNotForExpiredActions curEpoch votes =
+  checkDisallowedVotes votes canVoteOn VotingOnExpiredGovAction
+  where
+    canVoteOn _ GovActionState {gasExpiresAfter} = curEpoch <= gasExpiresAfter
 
 checkVotersAreValid ::
   forall era.
   ConwayEraPParams era =>
   CommitteeState era ->
   [(Voter (EraCrypto era), GovActionState era)] ->
   Test (ConwayGovPredFailure era)
-checkVotersAreValid committeeState voters =
-  let canVoteOn voter govAction = case voter of
-        CommitteeVoter {} -> isCommitteeVotingAllowed committeeState govAction
-        DRepVoter {} -> isDRepVotingAllowed govAction
-        StakePoolVoter {} -> isStakePoolVotingAllowed govAction
-      disallowedVoters =
-        filter
-          (\(voter, action) -> not $ voter `canVoteOn` gasAction action)
-          voters
-   in failureOnNonEmpty disallowedVoters $
-        DisallowedVoters . fmap (second gasId)
+checkVotersAreValid committeeState votes =
+  checkDisallowedVotes votes canVoteOn DisallowedVoters
+  where
+    canVoteOn voter gas = case voter of
+      CommitteeVoter {} -> isCommitteeVotingAllowed committeeState (gasAction gas)
+      DRepVoter {} -> isDRepVotingAllowed (gasAction gas)
+      StakePoolVoter {} -> isStakePoolVotingAllowed (gasAction gas)
+
+checkBootstrapVotes ::
+  forall era.
+  EraPParams era =>
+  PParams era ->
+  [(Voter (EraCrypto era), GovActionState era)] ->
+  Test (ConwayGovPredFailure era)
+checkBootstrapVotes pp votes
+  | HF.bootstrap (pp ^. ppProtocolVersionL) =
+      checkDisallowedVotes votes (flip canVoteOn) DisallowedVotesDuringBootstrap
+  | otherwise = pure ()
+  where
+    canVoteOn gas =
+      \case
+        DRepVoter {} | gasAction gas == InfoAction -> True
+        _ -> isBootstrapAction $ gasAction gas
```

Comment:
I recommend changing the order of arguments in `checkDisallowedVotes` in such a way that this is possible:
```suggestion
      checkDisallowedVotes votes DisallowedVotesDuringBootstrap $ \ gas ->
        \case
          DRepVoter {} | gasAction gas == InfoAction -> True
          _ -> isBootstrapAction $ gasAction gas
  | otherwise = pure ()
```

Note swap of arguments of predicate failure and `canVoteOn`, which allows it to become a lambda.

Also swapping arguments of `canVoteOn` is useful for lambda case (the reason why you used flip). This flipping of arguments sort of makes the function read `canBeVotedOnBy` :slightly_smiling_face: 

Same transformation can be then done to `checkVotersAreValid` and `checkVotesAreNotForExpiredActions`

Reply:
That's wonderful, thank you!